osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs

```diff
@@ -327,10 +327,29 @@ private void updateBeatmapAvailability(ValueChangedEvent<BeatmapAvailability> av
                 if (client.LocalUser?.State == MultiplayerUserState.Ready)
                     client.ChangeState(MultiplayerUserState.Idle);
             }
-            else
+
+            UpdateWorkingBeatmap();
```

Comment:
Having trouble understanding this. I think the part which is relevant to be called here should be extracted to a private method and called directly. If there's a reason for completely refreshing the working beatmap in this specific flow (ie. is it for some reason not handled by `beatmapUpdated` in the base class event flow?) it needs an inline comment.

Reply:
I'm not surprised, this is all pretty bad.

The biggest issue here was that `RoomSubScreen` was relying on events from `BeatmapManager` to update the working beatmap, rather than using the `BeatmapAvailabilityTracker`. This was causing scenarios where the availability tracker fired its event to let the `MultiplayerMatchSubScreen` know the beatmap's come available _before_ the scheduled callback from the `BeatmapManager` fired to change the working beatmap in `RoomSubScreen`.

Now that I've thought about this again, the easiest solution is to depend on the availability tracker inside `RoomSubScreen` also, which resulted in being able to revert all other changes.

I still liked the `GetWorkingBeatmap(null)` case I had in `UpdateWorkingBeatmap`, just to show to the user that they don't, in-fact, have the beatmap. `GetWorkingBeatmap()` will return a valid `WorkingBeatmap` regardless of whether it's in a soft-deleted state. This can be done in a separate PR, though.

So the sequence of events is:
1. `MultiplayerMatchSubScreen` changes the selected item
2. `RoomSubScreen.selectedItemChanged()` is called which then calls `RoomSubScreen.updateWorkingBeatmap()`
3. After the beatmap becomes available, `BeatmapAvailabilityTracker` starts firing off its event.
4. The event calls `RoomSubScreen.updateWorkingBeatmap` is first, updating the global `Beatmap` bindable. Then, `MultiplayerMatchSubScreen.updateBeatmapAvailability` is called which may attempt to enter spectate mode as a result of the beatmap becoming available.

It's still a bit of a mess but the flow of events is a bit more manageable now that `BeatmapManager` isn't in the mix.

Reply:
tl;dr: My recent commit reverted these changes and applied a simpler fix.